lms/djangoapps/edxnotes/helpers.py

```diff
@@ -0,0 +1,349 @@
+    """
+    # pylint: disable=method-hidden
+    def default(self, obj):
+        if isinstance(obj, datetime):
+            return get_default_time_display(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def get_id_token(user):
+    """
+    Generates JWT ID-Token, using or creating user's OAuth access token.
+    """
+    try:
+        client = Client.objects.get(name="edx-notes")
+    except Client.DoesNotExist:
+        raise ImproperlyConfigured("OAuth2 Client with name 'edx-notes' is not present in the DB")
+    try:
+        access_token = AccessToken.objects.get(
+            client=client,
+            user=user,
+            expires__gt=now()
+        )
+    except AccessToken.DoesNotExist:
+        access_token = AccessToken(client=client, user=user)
+        access_token.save()
+
+    id_token = oidc.id_token(access_token)
+    secret = id_token.access_token.client.client_secret
+    return id_token.encode(secret)
+
+
+def get_token_url(course_id):
+    """
+    Returns token url for the course.
+    """
+    return reverse("get_token", kwargs={
+        "course_id": course_id.to_deprecated_string(),
+    })
+
+
+def send_request(user, course_id, path="", query_string=""):
+    """
+    Sends a request with appropriate parameters and headers.
+    """
+    url = get_endpoint(path)
+    params = {
+        "user": anonymous_id_for_user(user, None),
+        "course_id": unicode(course_id).encode("utf-8"),
+    }
+
+    if query_string:
+        params.update({
+            "text": query_string,
+        })
+
+    try:
+        response = requests.get(
+            url,
+            headers={
+                "x-annotator-auth-token": get_id_token(user)
+            },
+            params=params
+        )
+    except RequestException:
+        raise EdxNotesServiceUnavailable(_("EdxNotes Service is unavailable. Please try again in a few minutes."))
+
+    return response
+
+
+def get_parent_unit(xblock):
+    """
+    Find vertical that is a unit, not just some container.
+    """
+    while xblock:
+        xblock = xblock.get_parent()
+        if xblock is None:
+            return None
+        parent = xblock.get_parent()
+        if parent is None:
+            return None
+        if parent.category == 'sequential':
+            return xblock
+
+
+def preprocess_collection(user, course, collection):
+    """
+    Reprocess provided `collection(list)`: adds information about ancestor,
+    converts "updated" date, sorts the collection in descending order.
+
+    Raises:
+        ItemNotFoundError - when appropriate module is not found.
+    """
+
+    store = modulestore()
+    filtered_collection = list()
+    cache = {}
+    with store.bulk_operations(course.id):
+        for model in collection:
+            model.update({
+                u"text": markupsafe.escape(model["text"]),
+                u"quote": markupsafe.escape(model["quote"]),
```

Comment:
> What is quote?

part of the text which has drawn student's attention. Highlighted one. Text is student's reflections on that.